instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/EvRerouteTest.kt

```diff
@@ -0,0 +1,217 @@
+import com.mapbox.navigation.core.MapboxNavigationProvider
+import com.mapbox.navigation.instrumentation_tests.R
+import com.mapbox.navigation.instrumentation_tests.activity.EmptyTestActivity
+import com.mapbox.navigation.instrumentation_tests.utils.MapboxNavigationRule
+import com.mapbox.navigation.instrumentation_tests.utils.assertions.RouteProgressStateTransitionAssertion
+import com.mapbox.navigation.instrumentation_tests.utils.history.MapboxHistoryTestRule
+import com.mapbox.navigation.instrumentation_tests.utils.http.MockDirectionsRequestHandler
+import com.mapbox.navigation.instrumentation_tests.utils.idling.RouteProgressStateIdlingResource
+import com.mapbox.navigation.instrumentation_tests.utils.location.MockLocationReplayerRule
+import com.mapbox.navigation.instrumentation_tests.utils.readRawFileText
+import com.mapbox.navigation.instrumentation_tests.utils.routes.RoutesProvider
+import com.mapbox.navigation.testing.ui.BaseTest
+import com.mapbox.navigation.testing.ui.utils.getMapboxAccessTokenFromResources
+import com.mapbox.navigation.testing.ui.utils.runOnMainSync
+import com.mapbox.navigation.utils.internal.logE
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import java.util.concurrent.CountDownLatch
+
+class EvRerouteTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::class.java) {
+
+    @get:Rule
+    val mapboxNavigationRule = MapboxNavigationRule()
+
+    @get:Rule
+    val mockLocationReplayerRule = MockLocationReplayerRule(mockLocationUpdatesRule)
+
+    @get:Rule
+    val mapboxHistoryTestRule = MapboxHistoryTestRule()
+
+    private lateinit var mapboxNavigation: MapboxNavigation
+
+    private lateinit var offRouteIdlingResource: RouteProgressStateIdlingResource
+    private lateinit var locationTrackingIdlingResource: RouteProgressStateIdlingResource
+
+    @Before
+    fun setup() {
+        Espresso.onIdle()
+
+        runOnMainSync {
+            mapboxNavigation = MapboxNavigationProvider.create(
+                NavigationOptions.Builder(activity)
+                    .accessToken(getMapboxAccessTokenFromResources(activity))
+                    .build()
+            )
+            mapboxHistoryTestRule.historyRecorder = mapboxNavigation.historyRecorder
+        }
+        locationTrackingIdlingResource = RouteProgressStateIdlingResource(
+            mapboxNavigation,
+            RouteProgressState.TRACKING
+        )
+        offRouteIdlingResource = RouteProgressStateIdlingResource(
+            mapboxNavigation,
+            RouteProgressState.OFF_ROUTE
+        )
+    }
+
+    override fun setupMockLocation(): Location {
+        val mockRoute = RoutesProvider.california_ev_route(activity)
+        return mockLocationUpdatesRule.generateLocationUpdate {
+            latitude = mockRoute.routeWaypoints.first().latitude()
+            longitude = mockRoute.routeWaypoints.first().longitude()
+        }
+    }
+
+    @Test
+    fun ev_reroute() {
```

Comment:
When I run this test locally, it sometimes fails because off route was not detected. It might be because of real location updates.
If we leave it as is, the test might become flaky.

Here's the error:
```
java.lang.RuntimeException: java.util.concurrent.ExecutionException: androidx.test.espresso.IdlingResourceTimeoutException: Wait for [RouteProgressStateIdlingResource] to become idle timed out
at androidx.test.espresso.Espresso.onIdle(Espresso.java:15)
at androidx.test.espresso.Espresso.onIdle(Espresso.java:1)
at com.mapbox.navigation.instrumentation_tests.core.EvRerouteTest$ev_reroute$4.invoke(EvRerouteTest.kt:177)
at com.mapbox.navigation.instrumentation_tests.core.EvRerouteTest$ev_reroute$4.invoke(EvRerouteTest.kt:176)
at com.mapbox.navigation.instrumentation_tests.utils.history.MapboxHistoryTestRule.stopRecordingOnCrash(MapboxHistoryTestRule.kt:71)
at com.mapbox.navigation.instrumentation_tests.core.EvRerouteTest.ev_reroute(EvRerouteTest.kt:176)
... 42 trimmed
Caused by: java.util.concurrent.ExecutionException: androidx.test.espresso.IdlingResourceTimeoutException: Wait for [RouteProgressStateIdlingResource] to become idle timed out
at java.util.concurrent.FutureTask.report(FutureTask.java:123)
at java.util.concurrent.FutureTask.get(FutureTask.java:193)
at androidx.test.espresso.Espresso.onIdle(Espresso.java:10)
... 48 more
Caused by: androidx.test.espresso.IdlingResourceTimeoutException: Wait for [RouteProgressStateIdlingResource] to become idle timed out
at androidx.test.espresso.IdlingPolicy.handleTimeout(IdlingPolicy.java:4)
at androidx.test.espresso.base.UiControllerImpl$5.resourcesHaveTimedOut(UiControllerImpl.java:1)
at androidx.test.espresso.base.IdlingResourceRegistry$Dispatcher.handleTimeout(IdlingResourceRegistry.java:4)
at androidx.test.espresso.base.IdlingResourceRegistry$Dispatcher.handleMessage(IdlingResourceRegistry.java:6)
at android.os.Handler.dispatchMessage(Handler.java:102)
at androidx.test.espresso.base.Interrogator.loopAndInterrogate(Interrogator.java:14)
at androidx.test.espresso.base.UiControllerImpl.loopUntil(UiControllerImpl.java:8)
at androidx.test.espresso.base.UiControllerImpl.loopMainThreadUntilIdle(UiControllerImpl.java:17)
at androidx.test.espresso.Espresso$1.run(Espresso.java:1)
at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:462)
at java.util.concurrent.FutureTask.run(FutureTask.java:266)
at android.os.Handler.handleCallback(Handler.java:938)
at android.os.Handler.dispatchMessage(Handler.java:99)
at android.os.Looper.loopOnce(Looper.java:201)
at android.os.Looper.loop(Looper.java:288)
at android.app.ActivityThread.main(ActivityThread.java:7839)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
```